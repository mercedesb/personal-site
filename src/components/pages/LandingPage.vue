<template>
  <div>
    <PageHeader
      :color="page.color || 'brown'"
      :title="page.title"
      :preamble="page.preamble"
      :media="iconUrl"></PageHeader>
    <main v-if="page.mainContent" class='FlexContainer FlexContainer--justifyCenter'>
      <p class="PageContent">
        <parse-markdown :source="page.mainContent" />
      </p>
    </main>
    <div v-if="ctaLinks.length" class='FlexContainer FlexContainer--justifyCenter'>
      <div class='PageContent PageContent--wide FlexContainer FlexContainer--wrap'>
      <CTALink
        v-for="ctaLink in ctaLinks"
        :key="ctaLink.id"
        v-bind="ctaLink"
      />
    </div>
    </div>
  </div>
</template>

<script>
import PageHeader from '../PageHeader.vue'
import CTALink from '../CTALink.vue'

export default {
  components: {
    PageHeader, CTALink
  },
  computed: {
    page() {
      return this.$store.state.landingPage
    },
    iconUrl() {
      return this.getImageUrl(this.page.icon)
    },
    ctaLinks() {
      if (!this.page.ctaLinks) return []
      return this.page.ctaLinks.map((cta) => {
        return {
          id: cta.sys.id,
          color: this.page.color,
          title: cta.fields.title,
          icon: this.getImageUrl(cta.fields.icon),
          url: cta.fields.externalLink || cta.fields.internalLink.fields.urlSegment,
          external: !!cta.fields.externalLink
        }
      })
    }
  },
  created () {
    this.$store.dispatch('getLandingPage', this.$route.params.urlSegment)
  }
}
</script>

<style lang="scss">
</style>