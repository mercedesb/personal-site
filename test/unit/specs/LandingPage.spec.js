import LandingPage from '@/components/pages/LandingPage.vue'
import { TestUtility } from './TestUtility'

describe('LandingPage', () => {
  let component

  const initialProps = {
    urlSegment: 'about'
  }

  const store = {
    state: {
      landingPage: {
        ...TestUtility.landingPages[initialProps.urlSegment]
      }
    },
    actions: {
      getLandingPage: jest.fn()
    }
  }

  const shallow = propsData => TestUtility.shallow(LandingPage, {
    store,
    propsData: {
      ...initialProps,
      ...propsData
    }
  })

  describe('Snapshots', () => {
    it('matches snapshot', () => {
      component = shallow()
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('Properties', () => {
    it('has a urlSegment property', () => {
      component = shallow()
      expect(component.props().urlSegment).toBe('about')
    })
  })

  describe('Computed', () => {
    describe('page', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has page populated', () => {
        expect(component.vm.page).toBeDefined()
      })
    })

    describe('iconUrl', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has iconUrl populated', () => {
        expect(component.vm.iconUrl).toBeDefined()
      })
    })

    describe('ctaLinks', () => {
      describe('with ctaLinks in cms', () => {
        it('has ctaLinks populated', () => {
          component = shallow()
          expect(component.vm.ctaLinks).toBeDefined()
        })

        it('creates objects of the desired shape', () => {
          component = shallow()
          const expectedShape = {
            id: expect.any(String),
            color: expect.any(String),
            title: expect.any(String),
            icon: expect.any(String),
            url: expect.any(String),
            external: expect.any(Boolean)
          }

          expect(component.vm.ctaLinks[0]).toMatchObject(expectedShape)
        })
      })

      describe('without ctaLinks in cms', () => {
        it('has ctaLinks populated', () => {
          const localPage = {
            ...store.state.landingPage,
            ctaLinks: undefined
          }
          const localStore = {
            state: {
              landingPage: localPage
            },
            actions: store.actions
          }
          component = TestUtility.shallow(LandingPage, {
            store: localStore,
            propsData: {
              ...initialProps
            }
          })

          expect(component.vm.ctaLinks).toBeDefined()
        })
      })
    })
  })

  describe('Lifecycle', () => {
    describe('created', () => {
      it('dispatches getLandingPage to the store', () => {
        component = TestUtility.mount(LandingPage, {
          store,
          propsData: {
            ...initialProps
          }
        })
        expect(component.vm.$store.dispatch).toHaveBeenCalledWith('getLandingPage', initialProps.urlSegment)
      })
    })
  })
})
