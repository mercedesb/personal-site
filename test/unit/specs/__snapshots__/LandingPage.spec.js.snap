// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingPage Snapshots matches snapshot 1`] = `
<main
  class="LandingPage"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="FlexContainer PageContent PageContent--wide LandingPage-ctaContainer"
  >
    <!---->
    <!---->
  </div>
</main>
`;
